src/init.c

```diff
@@ -283,8 +306,16 @@ shmem_internal_init(int tl_requested, int *tl_provided)
 
     /* Find symmetric data */
 #ifdef __APPLE__
-    shmem_internal_data_base = (void*) get_etext();
-    shmem_internal_data_length = get_end() - get_etext();
+    size_t region_size;
+    static int in_bss;
+
+    /* Start of this memory region (at address 0) */
+    vm_address_t addr = get_base_address(mach_task_self(), &region_size, 0);
+    shmem_internal_data_base = (void*)addr;
+
+    /* End of the BSS memory region (using an unititialized static variable) */
+    addr = get_base_address(mach_task_self(), &region_size, &in_bss);
+    shmem_internal_data_length = ((void*)addr + region_size) - shmem_internal_data_base;
```

Comment:
Confused here, may be misunderstanding. Shouldn't we subtract the `region_size` here?

Reply:
The highest address of the BSS is `addr + region_size`, so to calculate the size of the whole data segment we subtract that highest address from the base address of the data segment (at address 0).  Does that make sense?

Is it confusing because I reuse the `addr` and `region_size` variables?  At this line, `addr` points to the start of the BSS and `region_size` is the size of the BSS.  I can use more variables for clarity if you'd like?

Reply:
Thanks @davidozog for the explanation and a bit of experiment helped me to understand here. Looks good.